hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/connector/ReadCacheP.java

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.connector;
+
+import com.hazelcast.cache.impl.CacheProxy;
+import com.hazelcast.client.cache.impl.ClientCacheProxy;
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.client.impl.HazelcastClientProxy;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.Partition;
+import com.hazelcast.jet.AbstractProcessor;
+import com.hazelcast.jet.Processor;
+import com.hazelcast.jet.ProcessorMetaSupplier;
+import com.hazelcast.jet.ProcessorSupplier;
+import com.hazelcast.jet.Processors;
+import com.hazelcast.jet.Traverser;
+import com.hazelcast.jet.impl.util.CircularListCursor;
+import com.hazelcast.nio.Address;
+
+import javax.annotation.Nonnull;
+import javax.cache.Cache;
+import java.util.AbstractMap;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import static com.hazelcast.client.HazelcastClient.newHazelcastClient;
+import static com.hazelcast.jet.Util.entry;
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.IntStream.range;
+
+public class ReadCacheP extends AbstractProcessor {
```

Comment:
could we extract Cache and Map into a common processor maybe? I see a LOT of duplication in the code